src/components/TVChartContainer/TVChartContainer.tsx

```diff
@@ -0,0 +1,142 @@
+import { useLocalStorage, useMedia } from "react-use";
+import { defaultChartProps, disabledFeaturesOnMonile, SaveLoadAdapter } from "./constants";
+import useTVDatafeed from "domain/tradingview/useTVDatafeed";
+import { IChartingLibraryWidget, IPositionLineAdapter } from "./charting_library/charting_library";
+import { getPeriodFromResolutions, supportedResolutions } from "domain/tradingview/helper";
+
+type Props = {
+  symbol: string;
+  chainId: number;
+  savedShouldShowPositionLines: boolean;
+  currentPositions: any[];
+  currentOrders: any[];
+  onSelectToken: () => void;
+  period: string;
+  setPeriod: (period: string) => void;
+};
+
+export default function TVChartContainer({
+  symbol,
+  chainId,
+  savedShouldShowPositionLines,
+  currentPositions,
+  currentOrders,
+  onSelectToken,
+  period,
+  setPeriod,
+}: Props) {
+  const chartContainerRef = useRef<HTMLDivElement | null>(null);
+  const tvWidgetRef = useRef<IChartingLibraryWidget | null>(null);
+  const [chartReady, setChartReady] = useState(false);
+  let [tvCharts, setTvCharts] = useLocalStorage(TV_SAVE_LOAD_CHARTS, []);
+  const datafeed = useTVDatafeed();
+  const isMobile = useMedia("(max-width: 550px)");
+
+  const drawLineOnChart = useCallback(
+    (title, price) => {
+      if (chartReady && tvWidgetRef.current?.activeChart?.().dataReady()) {
+        return tvWidgetRef.current
+          .activeChart()
+          .createPositionLine({ disableUndo: true })
+          .setText(title)
+          .setPrice(price)
+          .setQuantity("")
+          .setLineStyle(1)
+          .setLineLength(1)
+          .setBodyFont(`normal 12pt "Relative", sans-serif`)
+          .setBodyTextColor("#fff")
+          .setLineColor("#3a3e5e")
+          .setBodyBackgroundColor("#3a3e5e")
+          .setBodyBorderColor("#3a3e5e");
+      }
+    },
+    [chartReady]
+  );
+
+  useEffect(() => {
+    const lines: (IPositionLineAdapter | undefined)[] = [];
+    if (savedShouldShowPositionLines) {
+      currentPositions.forEach((position) => {
+        const { open, liq } = position;
+        lines.push(drawLineOnChart(open.title, open.price));
+        lines.push(drawLineOnChart(liq.title, liq.price));
+      });
+      currentOrders.forEach((order) => {
+        lines.push(drawLineOnChart(order.title, order.price));
+      });
+    }
+    return () => {
+      lines.forEach((line) => line?.remove());
+    };
+  }, [currentPositions, savedShouldShowPositionLines, currentOrders, drawLineOnChart]);
+
+  useEffect(() => {
+    if (chartReady && tvWidgetRef.current && symbol !== tvWidgetRef.current?.activeChart?.().symbol()) {
+      tvWidgetRef.current.setSymbol(symbol, tvWidgetRef.current.activeChart().resolution(), () => {});
+    }
+  }, [symbol, chartReady, period]);
+
+  useEffect(() => {
+    if (chartReady && tvWidgetRef.current) {
+      if (isMobile && tvWidgetRef.current.activeChart().getCheckableActionState("drawingToolbarAction")) {
```

Comment:
1. i think it's good to avoid nested `if`s
hook may start with `if (!chartReady || !tvWidgetRef.current || !isMobile) { return }`

2. and the same question for `tvWidgetRef.current` – it's not part of deps

Reply:
chartReady is the state and it depends on the `tvWidgetRef` so using that instead.